src/utils.ts

```diff
@@ -1344,3 +1344,92 @@ export async function fileIsAccessible(fileName: string, mode?: number) {
     return false;
   }
 }
+
+/** @internal
+ * This class implements the RTT sampling logic specified for [CSOT](https://github.com/mongodb/specifications/blob/bbb335e60cd7ea1e0f7cd9a9443cb95fc9d3b64d/source/client-side-operations-timeout/client-side-operations-timeout.md#drivers-use-minimum-rtt-to-short-circuit-operations)
+ *
+ * This is implemented as a [circular buffer](https://en.wikipedia.org/wiki/Circular_buffer) keeping
+ * the most recent `windowSize` samples
+ * */
+export class RTTSampler {
+  /** Index of the next slot to be overwritten */
+  private writeIndex: number;
+  private _length: number;
+  private _rttSamples: Float64Array;
+
+  constructor(windowSize = 10) {
+    this._rttSamples = new Float64Array(windowSize);
+    this._length = 0;
+    this.writeIndex = 0;
+  }
+
+  /**
+   * Adds an rtt sample to the end of the circular buffer
+   * When `windowSize` samples have been collected, `addSample` overwrites the least recently added
+   * sample
+   */
+  addSample(sample: number) {
+    this._rttSamples[this.writeIndex++] = sample;
+    if (this._length < this._rttSamples.length) {
+      this._length++;
+    }
+
+    this.writeIndex %= this._rttSamples.length;
+  }
+
+  /**
+   * When \< 2 samples have been collected, returns 0
+   * Otherwise computes the minimum value samples contained in the buffer
+   */
+  min(): number {
+    if (this._length < 2) return 0;
+    let min = this._rttSamples[0];
+    for (let i = 1; i < this._length; i++) {
+      if (this._rttSamples[i] < min) min = this._rttSamples[i];
+    }
+
+    return min;
+  }
+
+  /**
+   * Returns mean of samples contained in the buffer
+   */
+  average(): number {
+    if (this._length === 0) return 0;
+    let sum = 0;
+    for (let i = 0; i < this._length; i++) {
+      sum += this._rttSamples[i];
+    }
+
+    return sum / this._length;
+  }
+
+  /**
+   * Returns most recently inserted element in the buffer
+   * Returns null if the buffer is empty
+   * */
+  get last(): number | null {
+    if (this._length === 0) return null;
+    return this._rttSamples[this.writeIndex === 0 ? this._length - 1 : this.writeIndex - 1];
+  }
+
+  /** @remarks Added for testing */
+  get length(): number {
+    return this._length;
+  }
```

Comment:
The tests can still access the `private` properties, I don't see the need for this indirection, and despite the comment this just makes length "public" again. If you want to avoid type errors in tests you can write `expect(rttSampler).to.have.property('length', 1);`